packages/eslint-plugin/src/rules/consistent-generic-constructors.ts

```diff
@@ -0,0 +1,103 @@
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import { createRule } from '../util';
+
+type MessageIds = 'preferTypeAnnotation' | 'preferConstructor';
+type Options = ['type-annotation' | 'constructor'];
+
+export default createRule<Options, MessageIds>({
+  name: 'consistent-generic-constructors',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Enforce specifying generic type arguments on type annotation or constructor name of a constructor call',
+      recommended: 'strict',
+    },
+    messages: {
+      preferTypeAnnotation:
+        'The generic type arguments should be specified as part of the type annotation.',
+      preferConstructor:
+        'The generic type arguments should be specified as part of the constructor type arguments.',
+    },
+    fixable: 'code',
+    schema: [
+      {
+        enum: ['type-annotation', 'constructor'],
+      },
+    ],
+  },
+  defaultOptions: ['constructor'],
+  create(context, [mode]) {
+    return {
+      VariableDeclarator(node): void {
+        const sourceCode = context.getSourceCode();
+        const lhs = node.id.typeAnnotation?.typeAnnotation;
+        const rhs = node.init;
+        if (
+          !rhs ||
+          rhs.type !== AST_NODE_TYPES.NewExpression ||
+          rhs.callee.type !== AST_NODE_TYPES.Identifier
+        ) {
+          return;
+        }
+        if (
+          lhs &&
+          (lhs.type !== AST_NODE_TYPES.TSTypeReference ||
+            lhs.typeName.type !== AST_NODE_TYPES.Identifier ||
+            lhs.typeName.name !== rhs.callee.name)
+        ) {
+          return;
+        }
+        if (mode === 'type-annotation' && !lhs && rhs.typeParameters) {
+          const { typeParameters, callee } = rhs;
+          const typeAnnotation =
+            sourceCode.getText(callee) + sourceCode.getText(typeParameters);
+          context.report({
+            node,
+            messageId: 'preferTypeAnnotation',
+            fix(fixer) {
+              return [
+                fixer.remove(typeParameters),
+                fixer.insertTextAfter(node.id, ': ' + typeAnnotation),
+              ];
+            },
+          });
+        }
```

Comment:
nit, there's no need to even consider the other `if` if the mode doesn't match, so we can return early, maybe something like this?
```ts
if (mode === 'type-annotation') {
  if (!lhs && rhs.typeParameters) {
    // ...
  }
  return;
}
```